effectful/ops/types.py

```diff
@@ -15,40 +22,330 @@ class NotHandled(Exception):
+          This feature can be used to conveniently define the syntax of a
+          domain-specific language.
+
+        * Defining free variables:
+
+          Passing :func:`Operation.define` a type creates a free variable.
+
+          >>> from effectful.ops.semantics import evaluate
+          >>> x = Operation.define(int, name='x')
+          >>> y = x() + 1
+
+          ``y`` is free in ``x``, so it is not fully evaluated:
+
+          >>> print(str(y))
+          __add__(x(), 1)
+
+          We bind ``x`` by installing a handler for it:
+
+          >>> with handler({x: lambda: 2}):
+          ...     print(evaluate(y))
+          3
+
+          .. note::
+
+            Because the result of :func:`Operation.define` is always fresh, it's
+            important to be careful with variable identity.
+
+            Two operations with the same name that come from different calls to
+            ``Operation.define`` are not equal:
+
+            >>> x1 = Operation.define(int, name='x')
+            >>> x2 = Operation.define(int, name='x')
+            >>> x1 == x2
+            False
+
+            This means that to correctly bind a variable, you must use the same
+            operation object. In this example, ``scale`` returns a term with a
+            free variable ``x``:
+
+            >>> x = Operation.define(float, name='x')
+            >>> def scale(a: float) -> float:
+            ...     return x() * a
+
+            Binding the variable ``x`` as follows does not work:
+
+            >>> term = scale(3.0)
+            >>> fresh_x = Operation.define(float, name='x')
+            >>> with handler({fresh_x: lambda: 2.0}):
+            ...     print(str(evaluate(term)))
+            __mul__(x(), 3.0)
+
+            Only the original operation object will work:
+
+            >>> from effectful.ops.semantics import fvsof
+            >>> with handler({x: lambda: 2.0}):
+            ...     print(evaluate(term))
+            6.0
+
+        * Defining a fresh :class:`Operation`:
+
+          Passing :func:`Operation.define` an :class:`Operation` creates a fresh
+          operation with the same name and signature, but no default rule.
+
+          >>> fresh_select = Operation.define(select)
+          >>> print(str(fresh_select(1, 2)))
+          select(1, 2)
+
+          The new operation is distinct from the original:
+
+          >>> with handler({select: lambda x, y: y}):
+          ...     print(select(1, 2), fresh_select(1, 2))
+          2 select(1, 2)
+
+          >>> with handler({fresh_select: lambda x, y: y}):
+          ...     print(select(1, 2), fresh_select(1, 2))
+          1 2
+
+        """
         raise NotImplementedError
 
-    @abc.abstractmethod
-    def __default_rule__(self, *args: Q.args, **kwargs: Q.kwargs) -> Expr[V]:
+    @define.register(classmethod)
+    @classmethod
+    def _define_classmethod(cls, default, **kwargs):
+        return Operation._ClassMethodOpDescriptor(cls.define, default.__func__)
```

Comment:
This is an error, so how is this still working? Is the `classmethod` path in `define` not being exercised in tests?